crates/ra_assists/src/auto_import.rs

```diff
@@ -0,0 +1,780 @@
+use hir::db::HirDatabase;
+use ra_syntax::{
+    ast, AstNode, SyntaxNode, Direction, TextRange,
+    SyntaxKind::{ PATH, PATH_SEGMENT, COLONCOLON, COMMA }
+};
+use crate::assist_ctx::{AssistCtx, Assist, AssistBuilder};
+use itertools::{ Itertools, EitherOrBoth };
```

Comment:
An interesting next step along the lines of this PR is "optimize imports" action, which reformats imports, sorts `crate::` paths last and merges prefixes. 

Though perhaps to do that we'll need a "structured editing API", that is, something like "remove this import" rather than "replace this sub string with that sub string". There's no such API in ra_syntax yet, it would be fun to design and implement it.

Reply:
Maybe having something like a mutable AST should be enough?
```
let mut mtreelist = treelist.get_mutable();
mtreelist.add_path(MutPath.from_str("foo::bar"))?
indent(mtreelist);
let diffs = diff(trellis, mstreelist); // get text insertions/deletions
```
Apart from `ident` and `diff`, I thing everything else can be autogenerated.